src/infrastructure/amqp/AbstractConsumer.ts

```diff
@@ -0,0 +1,168 @@
+import type { Either } from '@lokalise/node-core'
+import type { Channel, Connection, Message } from 'amqplib'
+import type { ZodSchema } from 'zod'
+
+import type { Dependencies } from '../diConfig'
+import { globalLogger, resolveGlobalErrorLogObject } from '../errors/globalErrorHandler'
+
+import type { ConsumerErrorResolver } from './ConsumerErrorResolver'
+import type { CommonMessage } from './MessageTypes'
+import { AmqpMessageInvalidFormat, AmqpValidationError } from './amqpErrors'
+import { deserializeMessage } from './messageDeserializer'
+
+export interface Consumer {
+  consume(): void
+  close(): Promise<void>
+}
+
+const ABORT_EARLY_EITHER: Either<'abort', never> = {
+  error: 'abort',
+}
+
+export type ConsumerParams<MessagePayloadType extends CommonMessage> = {
+  queueName: string
+  messageSchema: ZodSchema<MessagePayloadType>
+}
+
+export abstract class AbstractConsumer<MessagePayloadType extends CommonMessage>
+  implements Consumer
+{
+  protected readonly queueName: string
+  protected readonly connection: Connection
+  // @ts-ignore
+  protected channel: Channel
+  protected readonly errorHandler: ConsumerErrorResolver
+  private readonly newRelicBackgroundTransactionManager
+  private isShuttingDown: boolean
+
+  private messageSchema: ZodSchema<MessagePayloadType>
+
+  constructor(
+    params: ConsumerParams<MessagePayloadType>,
+    { amqpConnection, consumerErrorResolver, newRelicBackgroundTransactionManager }: Dependencies,
+  ) {
+    this.connection = amqpConnection
+    this.errorHandler = consumerErrorResolver
+    this.isShuttingDown = false
+    this.queueName = params.queueName
+    this.messageSchema = params.messageSchema
+    this.newRelicBackgroundTransactionManager = newRelicBackgroundTransactionManager
+  }
+
+  private async init() {
+    this.isShuttingDown = false
+
+    // If channel exists, recreate it
+    if (this.channel) {
+      this.isShuttingDown = true
+      await this.destroyConnection()
+      this.isShuttingDown = false
+    }
+
+    this.channel = await this.connection.createChannel()
+    this.channel.on('close', () => {
+      if (!this.isShuttingDown) {
+        globalLogger.error(`AMQP connection lost!`)
+        this.init().catch((err) => {
+          globalLogger.error(err)
+          throw err
+        })
+      }
+    })
+    this.channel.on('error', (err) => {
+      const logObject = resolveGlobalErrorLogObject(err)
+      globalLogger.error(logObject)
+    })
+
+    await this.channel.assertQueue(this.queueName, {
+      exclusive: false,
+      durable: true,
+      autoDelete: false,
+    })
+  }
+
+  protected deserializeMessage(message: Message | null): Either<'abort', MessagePayloadType> {
```

Comment:
:large_orange_diamond: Why protected? Can't it's overridden by any child class.

Reply:
You are right, this was the case of premature openness. I was assuming that there _might_ be some cases when one class may want to deserialize differently, but then again - YAGNI, and since it's a template, it will be an easy change anyway. 

tl; dr: you are right, will change.